8290840: Refactor the "os" class

iklam opened the pull request:
Please see [JDK-8290840](https://bugs.openjdk.org/browse/JDK-8290840) for the detailed proposal.

The `os` class, declared in os.hpp, forms the major part of the HotSpot porting interface. Its structure has gradually deteriorated over the years as new ports are created and new APIs are added.

This RFE tries to address the following:

- Clearly specify where a porting API should be declared and defined among the various `os*.cpp` and `os*.hpp` files.
- Avoid the inappropriate inclusion of OS-specific APIs (such as the `os::Linux class`) by platform-independent source files.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290840](https://bugs.openjdk.org/browse/JDK-8290840): Refactor the "os" class


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [06ae96b9](https://git.openjdk.org/jdk/pull/9600/files/06ae96b98841049217df99afaeddd059de80d0f0)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to [11c3c3d3](https://git.openjdk.org/jdk/pull/9600/files/11c3c3d3cbc868ff4755923a02d1890acb830f6d)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [1c7f8ebe](https://git.openjdk.org/jdk/pull/9600/files/1c7f8ebe1a2c9f73a1bdb3b6abba7979466ce630)
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to [cf93fb74](https://git.openjdk.org/jdk/pull/9600/files/cf93fb743b15414f695b01649327759f57912290)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9600/head:pull/9600` \
`$ git checkout pull/9600`

Update a local copy of the PR: \
`$ git checkout pull/9600` \
`$ git pull https://git.openjdk.org/jdk pull/9600/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9600`

View PR using the GUI difftool: \
`$ git pr show -t 9600`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9600.diff">https://git.openjdk.org/jdk/pull/9600.diff</a>

</details>

dholmes-ora commented on src/hotspot/os/aix/os_aix.cpp:
```diff
@@ -2998,3 +2999,11 @@ bool os::supports_map_sync() {
 
 void os::print_memory_mappings(char* addr, size_t bytes, outputStream* st) {}
+
+void os::print_user_info(outputStream* st) {
+  os::Posix::print_user_info(st);
+}
+
+void os::print_active_locale(outputStream* st) {
+  os::Posix::print_active_locale(st);
+}
```
You should not need this indirection. The code in os::Posix::x should be the implementation os::x for the os class, but just defined in os_posix.cpp.

iklam commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -37,9 +37,73 @@
+// a part the HotSpot Porting APIs. They must be implemented in
+// the following files:
+//
+// - src/hotspot/os/<os>/os_<os>.inline.hpp
+// - src/hotspot/os_cpu/<os>_<cpu>/os_<os>_<cpu>.inline.hpp
```
I meant to say:

```
// ... must be implemented in the following *four* files files:

// - src/hotspot/os/<os>/os_<os>.inline.hpp
// - src/hotspot/os_cpu/<os>_<cpu>/os_<os>_<cpu>.inline.hpp
//    [explanation for the hpp files]
// - src/hotspot/os/<os>/os_<os>.cpp
// - src/hotspot/os_cpu/<os>_<cpu>/os_<os>_<cpu>.cpp
//    [explanation for the cpp files]
```

I'll change the order of the text to make it clear.

dholmes-ora commented on src/hotspot/os/posix/os_posix.cpp:
```diff
@@ -570,13 +570,13 @@ void os::Posix::print_user_info(outputStream* st) {
   ::umask(umsk);
   st->print("umask: %04o (", (unsigned) umsk);
-  print_umask(st, umsk);
+  os::Posix::print_umask(st, umsk);
```
There's really no reason `print_umask` has to be part of the os::Posix class, it could just be a static file function in os_posix.cpp. I suspect the same may be true for other things in os::Posix - I would think we only need things in os::Posix that are explicitly called from os::Linux or os::Bsd etc.

dholmes-ora reviewed (commented):
Updates look good thanks. Overall this seems good to me. I think there may be more refinement possible with regards to the os::XXX classes, but that can be future cleanup if you want to proceed with this version for now.

Thanks.

dholmes-ora reviewed (approved):
Mis-clicked. Approved.

gerard-ziemski commented on src/hotspot/share/runtime/os.inline.hpp:
```diff
@@ -27,6 +27,37 @@
+}
+#endif
+
+#ifndef HAVE_FUNCTION_DESCRIPTORS
+inline void* os::resolve_function_descriptor(void* p) {
+  return NULL;
+}
```
This is unnecessary and could be removed if did:

```
+#ifdef HAVE_FUNCTION_DESCRIPTORS
   // Used only on PPC.
   inline static void* resolve_function_descriptor(void* p);
+#endif
```

in `os.hpp`

iklam replied:
After this PR, os.hpp no longer includes any os- or cpu-specific os_xxx.hpp files. I did this to avoid including large amount of os-specific declarations (such as os::Linux, which is about 400 lines) into generic source files, which aren't interested in these declarations.

As a result of this change, there's no way to let os.hpp know whether `HAVE_FUNCTION_DESCRIPTORS` is defined for the current OS+CPU. If we want to do that, we need to either:

- include os_xxx.hpp back in os.hpp (which I want to avoid)
- add new header files such as `os_<os>.defs.hpp`  `os_<os>_<cpu>.defs.hpp` to define  `HAVE_FUNCTION_DESCRIPTORS`  (this will result in lots of header files that usually have nothing inside)

Since there are only 4 functions that fall under this pattern (functions that are used in generic code but most platforms have dummy implementations), I think what I have is a good compromise

gerard-ziemski replied:
Could we do:

```
#if defined(LINUX) && defined(PPC) && !defined(ABI_ELFv2)
  #define HAVE_FUNCTION_DESCRIPTORS
#endif
```

in os.hpp and whatever the equivalent for AIX/PPC is?

tstuefe replied:
```
#if defined(PPC) && !defined(ABI_ELFv2)
  #define HAVE_FUNCTION_DESCRIPTORS
#endif
```
should do the trick. AIX uses XCOFF, not Elf, so ABI_ELFv2 should never be defined.

But I have no emotions. We can leave it like this. Or do it in a follow up.

tstuefe replied:
```
#if defined(PPC) && !defined(ABI_ELFv2)
  #define HAVE_FUNCTION_DESCRIPTORS
#endif
```
should do the trick. AIX uses XCOFF, not Elf, so ABI_ELFv2 should never be defined.

But I have no emotions. We can leave it like this. Or do it in a follow up.

gerard-ziemski commented on src/hotspot/share/runtime/os.inline.hpp:
```diff
@@ -27,6 +27,37 @@
+}
+#endif
+
+#ifndef _WINDOWS
+// Currently used only on Windows.
+inline bool os::register_code_area(char *low, char *high) {
+  return true;
+}
+#endif
```
We could use some custom define, similar to HAVE_FUNCTION_DESCRIPTORS and remove this default impl just like with `resolve_function_descriptor` below?

gerard-ziemski reviewed (changes requested):
Looks good, however, I would prefer not to have to provide the default implementations for `register_code_area()` or `resolve_function_descriptor()`

tstuefe commented on src/hotspot/share/runtime/os.inline.hpp:
```diff
@@ -27,6 +27,37 @@
+// Below are inline functions that are rarely implemented by the platforms.
+// Provide default empty implementation.
+
+#ifndef PLATFORM_PRINT_NATIVE_STACK
```
Could you rename this to HAVE_PLATFORM_PRINT_NATIVE_STACK?

iklam replied:
Fixed

tstuefe commented on src/hotspot/os/aix/os_aix.inline.hpp:
```diff
@@ -31,6 +31,11 @@
+// Information about the protection of the page at address '0' on this os.
+inline bool os::zero_page_read_protected() {
+  return false;
+}
```
I have the feeling many of these functions are really just global constants and that the coding could be simplified further in followup RFEs.

iklam replied:
Yeah lets do more clean up in follow-up RFEs.

tstuefe commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -4489,7 +4489,7 @@ jint os::init_2(void) {
   // perform the workaround
   Linux::capture_initial_stack(JavaThread::stack_size_at_create());
-  workaround_expand_exec_shield_cs_limit();
+  Linux::workaround_expand_exec_shield_cs_limit();
```
You could move `workaround_expand_exec_shield_cs_limit()` into this file, declare it static, and remove all mentionings from headers. It now lives in os_linux_x86.cpp, but since it has to be wrapped with !zero !64bit already it is an odd misfit anyway. So its not much more "wrong" for it to live here.

iklam replied:
Fixed.

tstuefe reviewed (approved):
Very good work. Only small remarks/nitpickings.

One question though, how does it work if I want to use an API implemented on a per-platform base in shared code? E.g. `zero_page_read_protected()`? I include `os.hpp` to get the definition, but how do I get the implementation? I have to include the platform-specific inline header too? Still with the OS_HEADER macro?

Cheers, Thomas

iklam commented on src/hotspot/os/aix/os_aix.hpp:
```diff
@@ -177,6 +176,10 @@ class Aix {
   // Returns true if ok, false if error.
   static bool get_meminfo(meminfo_t* pmi);
+
+  // PPC only
```
Fixed

iklam commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -37,9 +37,73 @@
+//
+//   The Porting APIs declared as "inline" in os.hpp must be
+//   implemented in one of the two .inline.hpp files, depending on whether
+//   the feature is generic to the OS, or specific to a particular CPU
```
I clarified the comment:
```
//   The Porting APIs declared as "inline" in os.hpp must be
//   implemented in one of the two .inline.hpp files, depending on
//   whether the feature is specific to a particular CPU architecture
//   for this OS. 
```

iklam commented:
> Very good work. Only small remarks/nitpickings.
> 
> One question though, how does it work if I want to use an API implemented on a per-platform base in shared code? E.g. `zero_page_read_protected()`? I include `os.hpp` to get the definition, but how do I get the implementation? I have to include the platform-specific inline header too? Still with the OS_HEADER macro?
> 
> Cheers, Thomas

I think we should avoid creating non-portable os::xxx APIs and then call them from shared code. There were some calls to os::Posix::xxx from the shared code that I fixed in this PR.

If you really want to make a platform-specific call, you should just do this (sadly we have many such cases)

```
#ifdef LINUX
linux_blah();
#endif
```

Putting linux_blah() into the os class seems pointless.

tstuefe commented on src/hotspot/os/linux/os_linux.cpp:
```diff
@@ -4461,6 +4461,89 @@ void os::Linux::numa_init() {
+ * Affects IA32: RHEL 5 & 6, Ubuntu 10.04 (LTS), 10.10, 11.04, 11.10, 12.04.
+ * @see JDK-8023956
+ */
+void workaround_expand_exec_shield_cs_limit() {
```
Make static?

tstuefe reviewed (approved):
Still good.

iklam commented:
@tstuefe @gerard-ziemski 

I've simplified `os::resolve_function_descriptor()` according to your suggestions, but I changed the #ifdefs to be more exact:

```
  // Used only on PPC for AIX, or LINUX (ppc but not ppcle). 
  static void* resolve_function_descriptor(void* p)
#if defined(PPC) && (defined(AIX) || (defined(LINUX) && !defined(ABI_ELFv2)))
    ;
#else
  { return NULL; }
#endif
```

If we use "#if defined(PPC) && !defined(ABI_ELFv2)))" it would be very difficult to understand.

To be honest, I am not a big fan of adding #ifdefs into os.hpp just to save a few lines of code. The #ifdefs are hard to understand. They are also fragile if you need to add a new platform that also support this functionality.

I am (semi) OK with doing this for resolve_function_descriptor(), as that's a PPC-only thing that's unlikely to apply for other platforms. However, I don't want to use the same pattern for os::platform_print_native_stack() because we may want to implement this for other planforms.

dholmes-ora commented on src/hotspot/os_cpu/linux_ppc/os_linux_ppc.cpp:
```diff
@@ -517,8 +517,8 @@ int os::extra_bang_size_in_bytes() {
 
-#ifdef HAVE_FUNCTION_DESCRIPTORS
-void* os::Linux::resolve_function_descriptor(void* p) {
+#if !defined(ABI_ELFv2)
```
I don't think many people will be able to map this to any specific platform. At a minimum some comment is needed to explain this.

dholmes-ora reviewed (approved):
Still seems okay to me.

tstuefe commented:
> 
> I think we should avoid creating non-portable os::xxx APIs and then call them from shared code. There were some calls to os::Posix::xxx from the shared code that I fixed in this PR.

I don't think that's what I meant. I meant APIs that exist for all OSes but have different expressions on each OS. So the prototype should live in shared os. But I want to pull a different implementation for each OS. 

This only affects inline functions. If the different implementations live in os_xx.cpp it works automatically through link time resolution.

I see there are only a few examples for what I meant. Many inline functions return platform dependend values that those live in variables in shared code (e.g. os::vm_page_size(), ...). Others are just not inline (os.create_thread(), ...). Which leaves us with the likes of `zero_page_read_protected()` etc, and I see you put them into os.inline.hpp and there you include the OS specific headers too.

I guess this can mostly be avoided. E.g. there is zero reason why `zero_page_read_protected()` cannot be a static boolean variable in os. For future RFEs.

iklam commented:
> > I think we should avoid creating non-portable os::xxx APIs and then call them from shared code. There were some calls to os::Posix::xxx from the shared code that I fixed in this PR.
> 
> I don't think that's what I meant. I meant APIs that exist for all OSes but have different expressions on each OS. So the prototype should live in shared os. But I want to pull a different implementation for each OS.
> 
> This only affects inline functions. If the different implementations live in os_xx.cpp it works automatically through link time resolution.
> 
> I see there are only a few examples for what I meant. Many inline functions return platform dependend values and those live in variables in shared code (e.g. os::vm_page_size(), ...). Others are just not inline (os.create_thread(), ...). Which leaves us with the likes of `zero_page_read_protected()` etc, and I see you put them into os.inline.hpp and there you include the OS specific headers too.
> 
> I guess this can mostly be avoided. E.g. there is zero reason why `zero_page_read_protected()` cannot be a static boolean variable in os. For future RFEs.

Thanks for pointing this out. I've updated the comments in os.hpp to be more clear about the inline functions.

The design is -- if `os.hpp` declares an "inline" function without a body, the implementation must be  in `os_<os>.inline.hpp`  or `os_<os>_<cpu>.inline.hpp`.  It's not possible for some platforms to choose to implement such a function in a .cpp file.

Conversely, if a function is NOT declared "inline" in `os.hpp`, it must NOT be implemented in any of the `os*inline.hpp` files.

Hence, any source file that references this function must include `os.inline.hpp`.

Without such a rule, it's possible for the build to succeed on some platforms (especially if the compiler is more lenient on missing inline function definitions) and fail on others. As a result, it would be easier to break the repo as not everyone is able to build all platforms before integrating a PR.

This rule is similar to how we handle "inline" declarations in regular HotSpot headers.

iklam commented:
> I see there are only a few examples for what I meant. Many inline functions return platform dependend values and those live in variables in shared code (e.g. os::vm_page_size(), ...). Others are just not inline (os.create_thread(), ...). Which leaves us with the likes of `zero_page_read_protected()` etc, and I see you put them into os.inline.hpp and there you include the OS specific headers too.
> 
> I guess this can mostly be avoided. E.g. there is zero reason why `zero_page_read_protected()` cannot be a static boolean variable in os. For future RFEs.

Yes, this can certainly be improved in follow-up RFE.

stefank commented on src/hotspot/os_cpu/linux_zero/os_linux_zero.cpp:
```diff
@@ -33,6 +33,8 @@
 #include "interpreter/interpreter.hpp"
 #include "memory/allocation.inline.hpp"
 #include "nativeInst_zero.hpp"
+#include "os_posix.hpp"
+#include "os_linux.hpp"
```
Sort

stefank commented on src/hotspot/share/runtime/os.inline.hpp:
```diff
@@ -27,6 +27,31 @@
 
 #include "runtime/os.hpp"
 
+#include OS_HEADER(os)
 #include OS_HEADER_INLINE(os)
```
We usually only explicitly include .inline.hpp and don't and let that file bring in the corresponding .hpp file.

stefank commented on test/hotspot/gtest/runtime/test_os.cpp:
```diff
@@ -35,6 +35,10 @@
 #include "runtime/frame.inline.hpp"
 #include "runtime/threads.hpp"
 
```
Remove blank line

stefank reviewed (changes requested):
I love seeing the os files being cleaned up and restructured.

I started to look at the patch and found a significant number of nits that I think would be good to fix.

gerard-ziemski commented:
> @tstuefe @gerard-ziemski
> 
> I've simplified `os::resolve_function_descriptor()` according to your suggestions, but I changed the #ifdefs to be more exact:
> 
> ```
>   // Used only on PPC for AIX, or LINUX (ppc but not ppcle). 
>   static void* resolve_function_descriptor(void* p)
> #if defined(PPC) && (defined(AIX) || (defined(LINUX) && !defined(ABI_ELFv2)))
>     ;
> #else
>   { return NULL; }
> #endif
> ```
> 
> If we use "#if defined(PPC) && !defined(ABI_ELFv2)))" it would be very difficult to understand.
> 
> To be honest, I am not a big fan of adding #ifdefs into os.hpp just to save a few lines of code. The #ifdefs are hard to understand. They are also fragile if you need to add a new platform that also support this functionality.
> 
> I am (semi) OK with doing this for resolve_function_descriptor(), as that's a PPC-only thing that's unlikely to apply for other platforms. However, I don't want to use the same pattern for os::platform_print_native_stack() because we may want to implement this for other planforms.

Thank you for trying. I am not completely happy about the new #ifdef and it doesn't look as clean to me as I imagined it would, so I am OK with you going with the original way it way. I guess, we could address it in a follow up if we come up with a nicer way of doing this...

iklam commented:
> > @tstuefe @gerard-ziemski
> > I've simplified `os::resolve_function_descriptor()` according to your suggestions, but I changed the #ifdefs to be more exact:
> > ```
> >   // Used only on PPC for AIX, or LINUX (ppc but not ppcle). 
> >   static void* resolve_function_descriptor(void* p)
> > #if defined(PPC) && (defined(AIX) || (defined(LINUX) && !defined(ABI_ELFv2)))
> >     ;
> > #else
> >   { return NULL; }
> > #endif
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If we use "#if defined(PPC) && !defined(ABI_ELFv2)))" it would be very difficult to understand.
> > To be honest, I am not a big fan of adding #ifdefs into os.hpp just to save a few lines of code. The #ifdefs are hard to understand. They are also fragile if you need to add a new platform that also support this functionality.
> > I am (semi) OK with doing this for resolve_function_descriptor(), as that's a PPC-only thing that's unlikely to apply for other platforms. However, I don't want to use the same pattern for os::platform_print_native_stack() because we may want to implement this for other planforms.
> 
> Thank you for trying. I am not completely happy about the new #ifdef and it doesn't look as clean to me as I imagined it would, so I am OK with you going with the original way it way. I guess, we could address it in a follow up if we come up with a nicer way of doing this...

Hi Gerard, thanks for the review. I've reverted the #ifdef change and gone back to the original way. See commit [cf93fb7](https://github.com/openjdk/jdk/pull/9600/commits/cf93fb743b15414f695b01649327759f57912290)

stefank commented on src/hotspot/os/bsd/os_bsd.inline.hpp:
```diff
@@ -25,11 +25,14 @@
 
-// os_bsd.hpp included by os.hpp
-
 #include "runtime/os.hpp"
+#include "os_bsd.hpp"
```
Change to:
```
#include "os_bsd.hpp"

#include "runtime/os.hpp"
```

stefank commented on src/hotspot/os/linux/os_linux.inline.hpp:
```diff
@@ -25,11 +25,14 @@
 
-// os_linux.hpp included by os.hpp
-
 #include "runtime/os.hpp"
+#include "os_linux.hpp"
 #include "os_posix.inline.hpp"
```
Change to:
```
#include "os_linux.hpp"

#include "runtime/os.hpp"
#include "os_posix.inline.hpp"
```

stefank commented on src/hotspot/os/windows/os_windows.inline.hpp:
```diff
@@ -27,10 +27,15 @@
 
 // os_windows.hpp included by os.hpp
 
+#include "os_windows.hpp"
 #include "runtime/javaThread.hpp"
 #include "runtime/mutex.hpp"
```
Change to:
```
#include "os_windows.hpp"

#include "runtime/javaThread.hpp"
#include "runtime/mutex.hpp"
```

stefank reviewed (approved):
Three more nits, otherwise looks good.

iklam commented:
Thanks @tstuefe @gerard-ziemski @stefank @dholmes-ora for the review.
Passed build tiers 1 and 5.
/integrate
